### DISCUSSION TITLE
Typo reveals hidden parameter in wb_add_fill

### DISCUSSION THREAD

**Discussion by Arthfael** [JanMarvin/openxlsx2#1099] [Q&A] [answered]
My issue is actually a double one. I am trying writing a table which contains text and quantitative data columns with some missing values. The table is formatted with banded rows, which I find useful for the text. I then apply over it conditional formatting to the quantitative data, but since NAs are not in the scale, they retain the banded background. To correct for this, before applying the conditional formatting I have tried to apply a white background on the whole quantitative data range. This is where the issues occur:

- First issue:
```
WorkBook <- wb_add_data_table(WorkBook, sheetnm, data,
                                                    dims = wb_dims(rows = 2, cols = 1),
                                                    col_names = TRUE, table_style = "TableStyleMedium2",
                                                    banded_rows = TRUE, banded_cols = FALSE)
dimRg <- wb_dims(rows = (1:nrow(data))+2, cols = blocks$Columns[[i]])
# (The table header is in row 2 so the data starts at row 3; blocks$Columns[[i]] corresponds to the indices of a block of consecutive quantitative columns in data)
WorkBook <- wb_add_fill(WorkBook,
                        sheetnm,
                        dimRg,
                        color = "white")
Error: Expecting a string vector: [type=NULL; required=STRSXP].
```
I can try to provide a reprex, but looking at my data all looks correct: this happens even if I write only the quantitative columns, a subset of 10 rows, all NaNs replaced with NA...
I am running on Windows. I looked up this error code and whilst it seems to crop up in multiple threads for different packages, I could not see an easy pattern to explain what is happening.

- The second issue is more of a weird observation:
Originally, I had made a small typo, writing "coor" instead of "color". The above error only occurred when I corrected the typo. 
```
WorkBook <- wb_add_fill(WorkBook
                        sheetnm,
                        dimRg,
                        coor = "white")
```
Surprisingly, using argument "coor" - which as far as I could find out is not documented anywhere - does not cause an error. Instead, I get more or less the expected outcome... except that the background of my #N/A cells is... yellow, not white!
This actually looks even better than my original plan, as they really stand out. So I will actually keep the typo in my code for now... but I would like to know what is happening and what argument coor actually is for.

**Comment by JanMarvin** [comment]
Hi @Arthfael , thanks for the question. There might be an issue with `wb_dims()` in release 1.8 see #1093. Could you check if `dimRg` returns the expected region? 

Regarding `wb_add_fill()`, there is no hidden argument. What you experience is a side effect of our attempt to provide backward compatible code. Coming from `openxlsx`, many arguments were camel case, but our functions were already snake case. When I began with `openxlsx2` I didn't care that much and we ended up with a strange mix of snake case and camel case. Since I already wrote a few examples on StackOverflow and didn't want those to become obsolete, I have added an internal conversion function that switches from camel case to snake case. So for example in `wb_add_fill()` both argument styles are accepted:

```R
wb_add_fill(gradient_fill = ...) # snake_case argument
wb_add_fill(gradientFill = ...)  # camelCase argument
```

Our conversion code converts `gradientFill` to `gradient_fill` and the user will not see an issue. To pass `gradientFill` the function takes arbitrary arguments as arguments with `...`. So `gradientFill` is passed from the wrapper function to R6 chain function `wb$add_fill()`. There it is converted and treated as if the user had passed `gradient_fill`. 

This has two side effects. The first is, that there is no additional check, that the argument matches specific arguments for a function. Therefore in your case `coor` is passed to the function, but completely ignored - because it does not match an argument. 
A second side effect is that partial matching isn't supported. In a plain R function you could pass `wb_add_fill(colo = ....)`. This wont work in `openxlsx2`, because `colo` is passed, but there is no partial match from `colo` to `color` and the argument will simply be ignored. 

I get that this is probably not as nice as the user expects, but it was a compromise to not create a lot of duplicated code (for snake case and camel case) and to avoid many internal checks (what arguments does the function accept?). At some point in the future, we might drop `camelCase` support (maybe with 2.0 and after a few months/quarters maybe a year (?) of throwing deprecation warnings), but for now that's the way it is. What you see right now is simply the default `color` from `wb_add_fill()`.

**Comment by Arthfael** [comment]
Dear Jan,
I just checked the output of wb_dims and it is correct:
> packageVersion("openxlsx2")
[1] ‘1.8’
> myRows <- (1:nrow(data))+2
> myCols <- blocks$Columns[[which(blocks$IsQuant)[1]]]
> min(myRows)
[1] 3
> max(myRows)
[1] 17109
> myCols
[1] 10 11 12 13 14 15 16
> dimRg <- wb_dims(rows = myRows, cols = myCols)
> dimRg
[1] "J3:P17109"
> WorkBook <- wb_add_fill(WorkBook, sheetnm, dimRg, color = "white")
Error: Expecting a string vector: [type=NULL; required=STRSXP].
> WorkBook <- wb_add_fill(WorkBook, sheetnm, dimRg, color = rgb(1, 1, 1))
Error: Expecting a string vector: [type=NULL; required=STRSXP].

Thanks for the clarification on arguments, this makes perfect sense - and also made me aware of the snake vs camel debate, which I had remained oblivious to until now - using inconsistently whichever method I fancied best at the time for my internal code.

Kind regards,

Armel

**Reply by JanMarvin** [reply, marked as answer]
Just realized that there might be another issue in your code. Colors require use of `wb_color()`. Could you please try the following:

```R
WorkBook <- wb_add_fill(wb = WorkBook, sheet = sheetnm, dims = dimRg, color = wb_color("white"))
WorkBook <- wb_add_fill(wb = WorkBook, sheet = sheetnm, dims = dimRg, color = wb_color(rgb(1,1,1)))
```

**Comment by Arthfael** [comment]
Woohooops... well that did the trick, thank you very much ^^ And thank you once again for a great package!
